tests/functional_tests/ffront_tests/test_single_static_assign.py

```diff
@@ -133,3 +133,513 @@ def test_empty_annotated_assign():
+                if True:
+                    a = a + 1
+                return a
+            """
+        )
+    )
+
+    expected = textwrap.dedent(
+        """
+        def f(a):
+            a__0 = a
+            if True:
+                a__1 = a__0 + 1
+            else:
+                a__1 = a__0
+            return a__1
+        """
+    ).strip()
+
+    assert result == expected
+
+
+def test_if_preservers_definite_assignment_analysis1():
+
+    result = ast.unparse(
+        ssaify_string(
+            """
+            def f():
+                if True:
+                    a = 1
+                else:
+                    b = 0
+                return a, b
+            """
+        )
+    )
+
+    expected = textwrap.dedent(
+        """
+        def f():
+            if True:
+                a__0 = 1
+            else:
+                b__0 = 0
+            return (a, b)
+        """
+    ).strip()
+
+    assert result == expected
+
+
+def test_if_preservers_definite_assignment_analysis2():
+
+    result = ast.unparse(
+        ssaify_string(
+            """
+            def f():
+                if True:
+                    a = 1
+                else:
+                    a = 0
+                return a, b
+            """
+        )
+    )
+
+    expected = textwrap.dedent(
+        """
+        def f():
+            if True:
+                a__0 = 1
+            else:
+                a__0 = 0
+            return (a__0, b)
+        """
+    ).strip()
+
+    assert result == expected
+
+
+def test_if_preservers_definite_assignment_analysis3():
+
+    result = ast.unparse(
+        ssaify_string(
+            """
+            def f():
+                a = 0
+                if True:
+                    a = 1
+                return a, b
+            """
+        )
+    )
+
+    expected = textwrap.dedent(
+        """
+        def f():
+            a__0 = 0
+            if True:
+                a__1 = 1
+            else:
+                a__1 = a__0
+            return (a__1, b)
+        """
+    ).strip()
+
+    assert result == expected
+
+
+def test_broken_collisions():
+    # Known bug of the current SSA implementation
+
+    result = ast.unparse(
+        ssaify_string(
+            """
+            a = a__0 + 1
+            return a, a__0
+            """
+        )
+    )
+
+    expected = textwrap.dedent(
+        """
+        a__0 = a__0 + 1
+        return (a__0, a__0)
+        """
+    ).strip()
+
+    assert result == expected
+
+
+def test_broken_collisions_function_parameters():
+    # Known bug of the current SSA implementation
+
+    result = ast.unparse(
+        ssaify_string(
+            """
+            def f(a, a__0):
+                return a__0
+            """
+        )
+    )
+
+    expected = textwrap.dedent(
+        """
+        def f(a, a__0):
+            a__0 = a
+            a__0__0 = a__0
+            return a__0__0
+        """
+    ).strip()
+
+    assert result == expected
```

Comment:
We could fix this with by using a tuple assign: https://github.com/BenWeber42/gt4py/commit/698f45b80c3301024fcb9aae19d19d1a7629f7e0

But I feel like the implementation gets quite complicated then...

Reply:
I think we should just use a common scheme for generated symbol names and include a check that verifies the user doesn't use such names himself. Something like `__a_0`.

Reply:
Clearly, there should be a way to generate _fresh_ variable names. But of course, we're still at python AST level for this pass, so it makes sense that this functionality doesn't exist at this stage. After symbol resolution, it would be easier to support this.

As for the scheme to create fresh variables, I usually prefer a proper scheme that doesn't restrict the user in any way. But there are easier schemes if you're ok restricting the user. A friend of mine just took the md5 sum of the current date/time and prepended it to fresh variable names with the reasoning that the probability of a collision is negligible :sweat_smile:

Reply:
Just repeating here the same suggestion I quickly discussed offline with @tehrengruber: we could use an easily distinguishable Unicode char from rare/unused alphabets to avoid collisions and, at the same time, keep the names short, intelligible and traceable (by _traceable_ I mean that it should be obvious to infer the original name of the symbols from their name mangled versions)

Reply:
This particular problem doesn't exist anymore. This PR uses the same name mangling scheme as the previous code. I think changing that is outside the scope of this PR.